### PR TITLE
TYP: define `subset` arg in `Styler`

### PR DESCRIPTION
**Pull request by attack68**
This works but someone with more knowledge might be able to point out a better definition of the type for `Subset`,

 i.e. here `Subset = Union[slice, Sequence, Index]`

It types all the `subset` args in `Styler`

### CODE REVIEW

**Comment by jreback** [comment]
cc @simonjayhawkins if any comments

File: pandas/io/formats/style_render.py [hunk 1]
```diff
@@ -809,7 +810,9 @@ def pred(part) -> bool:
             # slice(a, b, c)
             slice_ = [slice_]  # to tuplize later
     else:
-        slice_ = [part if pred(part) else [part] for part in slice_]
+        # error: Item "slice" of "Union[slice, Sequence[Any]]" has no attribute
+        # "__iter__" (not iterable) -> is specifically list_like in conditional
+        slice_ = [p if pred(p) else [p] for p in slice_]  # type: ignore[union-attr]
```
Comment (jreback) [inline pandas/io/formats/style_render.py:815]: if you return this directly it should work (as you are re-assigning to a different type)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style_render.py (hunk 1 above)
Comment (attack68) [reply pandas/io/formats/style_render.py:815]: it wasnt the reassignment it was the fact that slice_ is not always considered to be iterable inside the loop comprehension. Albeit in this clause we have already checked that it is `list_like` so is iterable here...

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @attack68 

if you can create a followon issue to try to integrate this to the rest of pandas; e.g. move Subset to _typing and then *use it* in .dropna, and so on (e.g. where ever we use a subset argument), may need to add a todo list

**Comment by simonjayhawkins** [comment]
> This works but someone with more knowledge might be able to point out a better definition of the type for `Subset`,

would/could/should this be the same as PositionalIndexer in pandas._typing

**Pull request merged**